src/main/java/org/kohsuke/github/GHCommitIterable.java

```diff
@@ -0,0 +1,72 @@
+package org.kohsuke.github;
+
+import java.util.Iterator;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Iterable for commit listing.
+ *
+ * @author Stephen Horgan
+ */
+class GHCommitIterable extends PagedIterable<GHCommit.File> {
+
+    private final GHRepository owner;
+    private final String sha;
+    private GHCommitFilesPage result;
+
+    /**
+     * Instantiates a new GH commit iterable.
+     *
+     * @param owner
+     *            the owner
+     * @param sha
+     *            the SHA of the commit
+     */
+    public GHCommitIterable(GHRepository owner, String sha) {
+        this.owner = owner;
+        this.sha = sha;
+    }
+
+    /**
+     * Iterator.
+     *
+     * @param pageSize
+     *            the page size
+     * @return the paged iterator
+     */
+    @Nonnull
+    @Override
+    public PagedIterator<GHCommit.File> _iterator(int pageSize) {
+
+        GitHubRequest request = owner.root().createRequest().withUrlPath(owner.getApiTailUrl("commits/" + sha)).build();
+
+        return new PagedIterator<>(
+                adapt(GitHubPageIterator.create(owner.root().getClient(), GHCommitFilesPage.class, request, pageSize)),
+                null);
+    }
```

Comment:
Move the logic that controls which iterating strategy to here. 
I haven't tried to compile this yet, so you may need to make some tweaks, but the basic idea should be there. 

You may or may not need to rerecord your test data. I think not....

```suggestion
    public PagedIterator<GHCommit.File> _iterator(int pageSize) {
        Iterator<GHCommit.File[]> pageIterator;
        if (results.getFiles().length() < GHCommit.GH_FILE_LIMIT_PER_COMMIT_PAGE) {
            // create a page iterator that only provides one page
            pageIterator = Collections.singleton(results.getFiles()).iterator();
        } else {
            // page size is controlled by the server for this iterator, do not allow it to be set by the caller
            pageSize = 0;

            GitHubRequest request = owner.root().createRequest().withUrlPath(owner.getApiTailUrl("commits/" + sha)).build();

           pageIterator = adapt(
                    GitHubPageIterator.create(owner.root().getClient(), GHCommitFilesPage.class, request, pageSize));
        }
        return new PagedIterator<>(pageIterator, null);
    }
```